src/cuda/cuda_topk_softmax.cuh

```diff
@@ -0,0 +1,136 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include <cfloat>
+#include <cmath>
+#include <cstdio>
+#include <cstdlib>
+#include <cub/cub.cuh>
+#include <limits>
+
+#include "cuda_topk.h"
+
+namespace Generators {
+namespace cuda {
+
+// Softmax for Sorted Input and k <= 256
+template <int kBlockSize, bool DoCopyIndices>
+__global__ void CopyAndSoftmaxKernel(int* final_indices, float* final_scores, const int* sorted_indices,
+                                     const float* sorted_scores, int k, float temperature, int input_stride) {
+  const int batch_idx = blockIdx.x;
+  const float* batch_sorted_scores = sorted_scores + batch_idx * input_stride;
+
+  // This implementation uses CUB for parallel reduction, which is efficient.
+  // The key fix is to correctly broadcast the reduction result (which CUB places in thread 0)
+  // to all other threads in the block using shared memory.
+  typedef cub::BlockReduce<float, kBlockSize> BlockReduce;
+  __shared__ typename BlockReduce::TempStorage temp_storage;
+  __shared__ float block_max_val;
+  __shared__ float block_sum_exp;
+
+  // STEP 1: Find max_val in parallel
+  // Each thread (where threadIdx.x < k) loads its score and applies temperature.
+  float thread_score =
+      (threadIdx.x < k) ? (batch_sorted_scores[threadIdx.x] / temperature) : -std::numeric_limits<float>::max();
+
+  // CUB reduces the values, placing the result in thread 0.
+  float max_val_reduced = BlockReduce(temp_storage).Reduce(thread_score, cub::Max());
```

Comment:
For algorithms like 'Selection sort', the top k values are bound to be in sorted order. Can there be an input flag to provide that hint so that we can avoid the search for max value ?

Reply:
Right. It is not needed. I've updated the code to use first elements for max value.